### PR TITLE
[rhi] Device upload readback functions

### PR DESCRIPTION
**Pull request by bobcao3**
Issue: #

### Brief Summary

Add device upload and readback functions, and use backend-specific fast path.

### CODE REVIEW

File: taichi/rhi/opengl/opengl_device.cpp [hunk 1]
```diff
@@ -546,10 +538,6 @@ DeviceAllocation GLDevice::allocate_memory(const AllocParams &params) {
     target_hint = GL_SHADER_STORAGE_BUFFER;
   } else if (params.usage && AllocUsage::Uniform) {
     target_hint = GL_UNIFORM_BUFFER;
-  } else if (params.host_write && params.host_read) {
```
Comment (ailzhang) [inline taichi/rhi/opengl/opengl_device.cpp:549]: nit: why do we remove these btw?

File: taichi/rhi/device.h [hunk 2]
```diff
@@ -678,6 +679,18 @@ class TI_DLL_EXPORT Device {
         this->allocate_memory(params));
   }
 
+  virtual RhiResult upload_data(DevicePtr *device_ptr,
```
Comment (ailzhang) [inline taichi/rhi/device.h:682]: nit: docstrings ;)?

**Review by ailzhang — APPROVED** [review]
Thanks!!

File: taichi/rhi/device.h (hunk 2 above)
Comment (bobcao3) [reply taichi/rhi/device.h:682]: Ah I forgot

File: taichi/rhi/opengl/opengl_device.cpp (hunk 1 above)
Comment (bobcao3) [reply taichi/rhi/opengl/opengl_device.cpp:549]: They don't actually modify the hint right now.

**Pull request merged**